src/group_theory/perm/concrete_cycle.lean

```diff
@@ -0,0 +1,543 @@
+    simp [hx] },
+  have hc : is_cycle (cycle_of p x) := is_cycle_cycle_of p hx,
+  rw nodup_iff_nth_le_inj,
+  rintros n m hn hm,
+  rw [length_to_list, ←order_of_is_cycle hc] at hm hn,
+  rw [←cycle_of_apply_self, ←ne.def, ←mem_support] at hx,
+  rw [nth_le_to_list, nth_le_to_list,
+      ←cycle_of_pow_apply_self p x n, ←cycle_of_pow_apply_self p x m],
+  cases n;
+  cases m,
+  { simp },
+  { rw [←hc.mem_support_pos_pow_iff_of_lt_order_of m.zero_lt_succ hm,
+        mem_support, cycle_of_pow_apply_self] at hx,
+    rw eq_comm,
+    simp [hx] },
+  { rw [←hc.mem_support_pos_pow_iff_of_lt_order_of n.zero_lt_succ hn,
+        mem_support, cycle_of_pow_apply_self] at hx,
+    simp [hx] },
+  intro h,
+  have hn' : ¬ order_of (p.cycle_of x) ∣ n.succ := nat.not_dvd_of_pos_of_lt n.zero_lt_succ hn,
+  have hm' : ¬ order_of (p.cycle_of x) ∣ m.succ := nat.not_dvd_of_pos_of_lt m.zero_lt_succ hm,
+  rw ←hc.support_pow_eq_iff at hn' hm',
+  rw [←nat.mod_eq_of_lt hn, ←nat.mod_eq_of_lt hm, ←pow_inj_mod],
+  refine support_congr _ _,
+  { rw [hm', hn'],
+    exact finset.subset.refl _ },
+  { rw hm',
+    intros y hy,
+    obtain ⟨k, rfl⟩ := hc.exists_pow_eq (mem_support.mp hx) (mem_support.mp hy),
+    rw [←mul_apply, (commute.pow_pow_self _ _ _).eq, mul_apply, h, ←mul_apply, ←mul_apply,
+        (commute.pow_pow_self _ _ _).eq] }
+end
+
+lemma next_to_list_eq_apply (p : perm α) (x y : α) (hy : y ∈ to_list p x) :
+  next (to_list p x) y hy = p y :=
+begin
+  rw mem_to_list_iff at hy,
+  obtain ⟨k, hk, hk'⟩ := hy.left.nat_of_mem_support _ hy.right,
+  rw ←nth_le_to_list p x k (by simpa using hk) at hk',
+  simp_rw ←hk',
+  rw [next_nth_le _ (nodup_to_list _ _), nth_le_to_list, nth_le_to_list, ←mul_apply, ←pow_succ,
+      length_to_list, pow_apply_eq_pow_mod_order_of_cycle_of_apply p (k + 1), order_of_is_cycle],
+  exact is_cycle_cycle_of _ (mem_support.mp hy.right)
+end
+
+lemma same_cycle.to_list_is_rotated {f : perm α} {x y : α} (h : same_cycle f x y) :
+  to_list f x ~r to_list f y :=
+begin
+  by_cases hx : x ∈ f.support,
+  { obtain ⟨_ | k, hk, hy⟩ := h.nat_of_mem_support _ hx,
+    { simp only [coe_one, id.def, pow_zero] at hy,
+      simp [hy] },
+    use k.succ,
+    apply list.ext_le,
+    { simp only [length_rotate, length_to_list, ←hy],
+      congr' 1,
+      ext z,
+      rw [mem_support, mem_support],
+      by_cases hz : same_cycle f x z,
+      { rw [hz.cycle_of_apply, same_cycle.cycle_of_apply],
+        simpa using hz },
+      { rw [cycle_of_apply_of_not_same_cycle hz, cycle_of_apply_of_not_same_cycle],
+        simpa using hz } },
+    { intros n hn hn',
+      suffices : (f ^ ((n + k.succ) % (f.cycle_of x).support.card)) x =
+        (f ^ n) ((f ^ k.succ) x),
+      { simpa [nth_le_rotate, nth_le_to_list, ←hy], },
+      rw [←cycle_of_pow_apply_self, ←mul_apply, ←pow_add, ←cycle_of_pow_apply_self f,
+          ←order_of_is_cycle, ←pow_eq_mod_order_of],
+      exact is_cycle_cycle_of _ (mem_support.mp hx) } },
```

Comment:
There are some good sublemmas to extract here:
 *  `f.to_list ((f ^ k) x) = (f.to_list x).rotate k`
 * `(f ^ (n % (f.cycle_of x).support.card)) x = (f ^ n) x`